### PR TITLE
Fix sighelp crash in linked files

### PR DESCRIPTION
**Pull request by rchande**
Fixes https://github.com/dotnet/roslyn/issues/14793#issuecomment-257662554
Tag @dotnet/roslyn-ide for review

### CODE REVIEW

**Comment by rchande** [comment]
@Pilchie We'll probably want to take this for RC.

**Comment by Pilchie** [comment]
@rchande Can you include the questions @MattGertz asked about?

@brettfo - can you do some buddy testing here?

**Comment by rchande** [comment]
@MattGertz Escrow shiproom info: 

* **Customer Scenario**: The user starts typing a tuple in a linked file. If tuple construction signature help is triggered, VS crashes.
* **Bugs**: https://github.com/dotnet/roslyn/issues/14793
* **Workarounds**: The only workarounds would be to use the option to disable signature help, or to not use tuples in a linked file
* **Risk**: The fix only affects signature help in linked files. With the change, linked file signature help guards against a particular cast failing. 
* **Performance Impact**: Low--no extra allocations or complexity changes
* **Regression**: This was introduced by https://github.com/dotnet/roslyn/pull/14645 during RC ask mode
* **RCA**: This particular bug was always lurking because of the unguarded cast in linked files signature help, however no testing was performed with tuples in linked files when adding the new signature help provider.
* **How was the bug found**: Dogfooding

**Comment by MattGertz** [comment]
@natidea Should we add this to today's list?

File: src/Features/CSharp/Portable/SignatureHelp/TupleConstructionSignatureHelpProvider.cs [hunk 1]
```diff
@@ -152,21 +152,30 @@ public override Boolean IsTriggerCharacter(Char ch)
...
-            var prefixParts = SpecializedCollections.SingletonEnumerable(new SymbolDisplayPart(SymbolDisplayPartKind.Punctuation, null, "("));
-            var suffixParts = SpecializedCollections.SingletonEnumerable(new SymbolDisplayPart(SymbolDisplayPartKind.Punctuation, null, ")"));
-            var separatorParts = GetSeparatorParts();
+            var prefixParts = SpecializedCollections.SingletonEnumerable(new SymbolDisplayPart(SymbolDisplayPartKind.Punctuation, null, "(")).ToTaggedText();
+            var suffixParts = SpecializedCollections.SingletonEnumerable(new SymbolDisplayPart(SymbolDisplayPartKind.Punctuation, null, ")")).ToTaggedText();
+            var separatorParts = GetSeparatorParts().ToTaggedText();
 
-            var items = tupleTypes.Select(t =>
-                new SignatureHelpItem(isVariadic: false,
+            List<SignatureHelpItem> items = tupleTypes.Select(tupleType => Convert(
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/SignatureHelp/TupleConstructionSignatureHelpProvider.cs]: var

**Comment by tmeschter** [comment]
LGTM

File: src/EditorFeatures/CSharpTest/SignatureHelp/TupleConstructionSignatureHelpProviderTests.cs [hunk 2]
```diff
@@ -200,5 +200,35 @@ static void Main(string[] args)
...
+        [WorkItem(14793, "https://github.com/dotnet/roslyn/issues/14793")]
+        [Fact, Trait(Traits.Feature, Traits.Features.SignatureHelp)]
+        public async Task DoNotCrashInLinkedFile()
+        {
+            var markup = @"<Workspace>
+    <Project Language=""C#"" CommonReferences=""true"" AssemblyName=""Proj1"" PreprocessorSymbols=""FOO"">
+        <Document FilePath=""SourceDocument""><![CDATA[
+class C
+{
+#if FOO
```
Comment (tmeschter) [inline src/EditorFeatures/CSharpTest/SignatureHelp/TupleConstructionSignatureHelpProviderTests.cs]: Is the use of a preprocessor directive significant?

**Pull request merged**